src/Microsoft.OData.Core/ODataNotificationReader.cs

```diff
@@ -113,18 +121,44 @@ public override Task<string> ReadToEndAsync()
         /// <param name="disposing">True if called from Dispose; false if called from the finalizer.</param>
         protected override void Dispose(bool disposing)
         {
-            if (disposing)
+            if (!this.disposed && disposing)
             {
-                if (this.listener != null)
+                // Tell the listener that the stream is being disposed.
+                if (synchronous)
                 {
-                    // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
-                    this.listener = null;
+                    this.listener?.StreamDisposed();
                 }
+                else
+                {
+                    this.listener?.StreamDisposedAsync().Wait();

```

Comment:
Out of curiosity, is it better to call a synchronous method or to `Wait()` on an async method?

Reply:
@habbes I have created a work item to do an investigation here and determine the better way